subprojects/core/src/main/java/org/gradle/execution/selection/DefaultBuildTaskSelector.java

```diff
@@ -61,6 +61,12 @@ public Filter resolveExcludedTaskName(BuildState defaultBuild, String taskName)
             // Too early to resolve excludes
             return new Filter(defaultBuild, Specs.satisfyNone());
         }
+        for (ProjectState project : defaultBuild.getProjects().getAllProjects()) {
```

Comment:
Could parallel `BuildState#isProjectsLoaded()` with a `BuildState.isProjectsCreated()`. 

Something like this:

```
@Override
    public boolean isProjectsCreated() {
        BuildProjectRegistry projectsForThisBuild = getProjectStateRegistry().findProjectsFor(getBuildIdentifier());
        return projectsForThisBuild != null && projectsForThisBuild.getAllProjects().stream().allMatch(ProjectState::isCreated);
    }
```

Didn't run any tests though.  Note that `getProjectStateRegistry().findProjectsFor(...)` returns `null` if project states are not available yet, whereas `getProjectStateRegistry().projectsFor(...)`  (called by`BuildState#getAllProjects()`) fails with an `IllegalStateException`.

Reply:
It seems that this new `isProjectsCreated` predicate could be used as the first guard in place of `isProjectsLoaded`, right?

Reply:
@bamboo points out that we should be able to check only the root project, as we do in the Configuratin Cache:

https://github.com/gradle/gradle/blob/93d70ab7fd76b71b116dc4237a15775f4cd5edac/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt#L857

That means we could have only:

```
     @Override
    public boolean isProjectsAvailable() {
        BuildProjectRegistry buildProjects = getProjectStateRegistry().findProjectsFor(getBuildIdentifier());
        return buildProjects != null && buildProjects.getRootProject().isCreated();
    }
```

and remove the extension method we have in the Configuration Cache.

Reply:
Also, we no longer need to check `isProjectsLoaded()` - as `isProjectsAvailable()` already does that check.